dsc/workflows/base/base_workflow.py

```diff
@@ -0,0 +1,199 @@
+            item_submission.upload_dspace_metadata(
+                self.s3_bucket, item_submission.metadata_keyname
+            )
+            sqs_client.send_submission_message(
+                item_submission.item_identifier,
+                self.workflow_name,
+                self.output_queue,
+                self.submission_system,
+                self.collection_handle,
+                item_submission.metadata_uri,
+                item_submission.bitstream_uris,
+            )
+        """
+
+    @final
+    def item_submissions_iter(self) -> Iterator[ItemSubmission]:
+        """Generate a batch of item submissions for the DSpace Submission Service.
+
+        MUST NOT be overridden by workflow subclasses.
+        """
+        for item_metadata in self.batch_metadata_iter():
+            item_identifier = self.get_item_identifier(item_metadata)
+            logger.info(f"Processing submission for '{item_identifier}'")
+            metadata_keyname = f"{self.s3_prefix}/{item_identifier}_metadata.json"
+            dspace_metadata = self.create_dspace_metadata(item_metadata)
+            if self.validate_dspace_metadata(dspace_metadata):
+                item_submission = ItemSubmission(
+                    dspace_metadata=dspace_metadata,
+                    bitstream_uris=self.get_bitstream_uris(item_identifier),
+                    metadata_keyname=metadata_keyname,
+                )
+                yield item_submission
+
+    @abstractmethod
+    def batch_metadata_iter(self) -> Iterator[dict[str, Any]]:
+        """Iterate through batch metadata to yield item metadata.
+
+        MUST be overridden by workflow subclasses.
+        """
+
+    @abstractmethod
+    def get_item_identifier(self, item_metadata: Any) -> str:  # noqa: ANN401
+        """Get identifier for an item submission according to the workflow subclass.
+
+        MUST be overridden by workflow subclasses.
+
+        Args:
+            item_metadata: The item metadata from which the item identifier is extracted.
+        """
+
+    @final
+    def create_dspace_metadata(self, item_metadata: dict[str, Any]) -> dict[str, Any]:
+        """Create DSpace metadata from the item's source metadata.
+
+        A metadata mapping is a dict with the format seen below:
+
+        {
+        "dc.contributor": {
+            "source_field_name": "contributor",
+            "language": None,
+            "delimiter": "|",
+        }
+
+        MUST NOT be overridden by workflow subclasses.
+
+        Args:
+            item_metadata: Item metadata from which the DSpace metadata will be derived.
+        """
+        metadata_entries = []
+        for field_name, field_mapping in self.metadata_mapping.items():
+            if field_name not in ["item_identifier", "source_system_identifier"]:
+
+                field_value = item_metadata.get(field_mapping["source_field_name"])
+                if field_value:
```

Comment:
I understand now that the mappings are complex objects and can have things like `delimiter` and `language`.

At a per workflow, per mapping level, is it worth considering adding an **optional** `required` property to the target field mapping?  Then you could do something like:

```python
if not field value and field_mapping.get("required",False):
  # raise an exception because an item didn't have a field that is required for this workflow/mapping
```

Reply:
Good call! `title` is the only required DSpace field but certain workflows may need others, I'll implement it for title at this point